src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+};
+
+use bitvec::field::BitField;
+use rand_core::RngCore;
+use std::io::Cursor;
+use subtle::{Choice, ConditionallyNegatable, ConditionallySelectable};
+
+use crate::{
+    codec::{decode_u32_items, encode_u32_items, CodecError, Decode, Encode},
+    field::FieldElement,
+    idpf::{
+        conditional_select_seed, conditional_swap_seed, conditional_xor_seeds, xor_seeds,
+        IdpfInput, IdpfValue,
+    },
+    vdaf::xof::{Seed as XofSeed, Xof, XofFixedKeyAes128, XofTurboShake128},
+};
+
+/// VIDPF-related errors.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum VidpfError {
+    /// Error when key's identifier are equal.
+    #[error("key's identifier should be different")]
+    SameKeyId,
+
+    /// Error when level does not fit in a 32-bit number.
+    #[error("level is not representable as a 32-bit integer")]
+    LevelTooBig,
+
+    /// Error during VIDPF evaluation: tried to access a level index out of bounds.
+    #[error("level index out of bounds")]
+    IndexLevel,
+
+    /// Error when weight's length mismatches the length in weight's parameter.
+    #[error("invalid weight length")]
+    InvalidWeightLength,
+
+    /// Failure when calling getrandom().
+    #[error("getrandom: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// [`VidpfInput`] represents the domain of an incremental point function.
+pub type VidpfInput = IdpfInput;
+
+/// [`VidpfWeight`] represents the codomain of an incremental point function.
+pub trait VidpfWeight: IdpfValue + for<'a> Add<&'a Self, Output = Self> {}
+
+/// [`Vidpf`] is a VIDPF instance.
+pub struct Vidpf<W: VidpfWeight, const NONCE_SIZE: usize> {
+    /// Any parameters required to instantiate a weight value.
+    weight_parameter: W::ValueParameter,
+}
+
+impl<W: VidpfWeight, const NONCE_SIZE: usize> Vidpf<W, NONCE_SIZE> {
+    /// Creates a VIDPF instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `weight_parameter`, any parameters required to instantiate a weight value.
+    pub const fn new(weight_parameter: W::ValueParameter) -> Self {
+        Self { weight_parameter }
+    }
+
+    /// The [`Vidpf::gen`] method splits an incremental point function `F` into two private keys
+    /// used by the aggregation servers, and a common public share.
+    ///
+    /// The incremental point function is defined as `F`: [`VidpfInput`] --> [`VidpfWeight`]
+    /// such that:
+    ///
+    /// ```txt
+    /// F(x) = weight, if x is a prefix of the input.
+    /// F(x) = 0,      if x is not a prefix of the input.
+    /// ```
+    ///
+    /// # Arguments
+    ///
+    /// * `input`, determines the input of the function.
+    /// * `weight`, determines the input's weight of the function.
+    /// * `nonce`, used to cryptographically bind some information.
+    pub fn gen(
+        &self,
+        input: &VidpfInput,
+        weight: &W,
+        nonce: &[u8; NONCE_SIZE],
+    ) -> Result<(Key, Key, PublicShare<W>), VidpfError> {
```

Comment:
nit: For symmetry with the `Vdaf` API, we might make the return type: `PublicShare<W>, [Key; 2]`